swarm/storage/mru/resource.go

```diff
@@ -122,114 +119,110 @@ type LookupParams struct {
-	startBlock uint64
-	lastPeriod uint32
-	lastKey    storage.Address
-	frequency  uint64
-	version    uint32
-	data       []byte
-	owner      common.Address
-	updated    time.Time
+	lastKey storage.Address
+	updated time.Time
 }
 
 // TODO Expire content after a defined period (to force resync)
 func (r *resource) isSynced() bool {
 	return !r.updated.IsZero()
 }
 
-func (r *resource) NameHash() common.Hash {
-	return r.nameHash
+func (r *resourceUpdate) Multihash() bool {
+	return r.multihash
 }
 
 // implements (which?) interface
 func (r *resource) Size(_ chan bool) (int64, error) {
 	if !r.isSynced() {
 		return 0, NewError(ErrNotSynced, "Not synced")
 	}
-	return int64(len(r.data)), nil
+	return int64(len(r.resourceUpdate.data)), nil
 }
 
 func (r *resource) Name() string {
 	return r.name
 }
 
-func (r *resource) UnmarshalBinary(data []byte) error {
-	r.startBlock = binary.LittleEndian.Uint64(data)
-	r.frequency = binary.LittleEndian.Uint64(data[8:])
-	copy(r.owner[:], data[16:16+common.AddressLength])
-	r.name = string(data[16+common.AddressLength:])
-	return nil
+type timestampProvider interface {
+	GetCurrentTime() uint64
 }
 
-func (r *resource) MarshalBinary() ([]byte, error) {
-	b := make([]byte, 16+len(r.name))
-	binary.LittleEndian.PutUint64(b, r.startBlock)
-	binary.LittleEndian.PutUint64(b[8:], r.frequency)
-	copy(b[16:], r.owner[:])
-	copy(b[16+common.AddressLength:], []byte(r.name))
-	return b, nil
+type DefaultTimestampProvider struct {
 }
 
-type headerGetter interface {
-	HeaderByNumber(context.Context, string, *big.Int) (*types.Header, error)
+func NewDefaultTimestampProvider() *DefaultTimestampProvider {
+	return &DefaultTimestampProvider{}
+}
+
+func (dtp *DefaultTimestampProvider) GetCurrentTime() uint64 {
+	return uint64(time.Now().Unix())
 }
 
 // Handler is the API for Mutable Resources
 // It enables creating, updating, syncing and retrieving resources and their update data
 type Handler struct {
-	chunkStore      *storage.NetStore
-	HashSize        int
-	signer          Signer
-	headerGetter    headerGetter
-	resources       map[string]*resource
-	hashPool        sync.Pool
-	resourceLock    sync.RWMutex
-	storeTimeout    time.Duration
-	queryMaxPeriods uint32
+	chunkStore        *storage.NetStore
+	HashSize          int
+	timestampProvider timestampProvider
+	resources         map[uint64]*resource
+	resourceLock      sync.RWMutex
+	storeTimeout      time.Duration
+	queryMaxPeriods   uint32
 }
 
 // HandlerParams pass parameters to the Handler constructor NewHandler
-// Signer and HeaderGetter are mandatory parameters
+// Signer and TimestampProvider are mandatory parameters
 type HandlerParams struct {
-	QueryMaxPeriods uint32
-	Signer          Signer
-	HeaderGetter    headerGetter
+	QueryMaxPeriods   uint32
+	TimestampProvider timestampProvider
+}
+
+var hashPool sync.Pool
+
+func init() {
+	hashPool = sync.Pool{
+		New: func() interface{} {
+			return storage.MakeHashFunc(resourceHashAlgorithm)()
+		},
+	}
```

Comment:
Moved the hashPool as a shared singleton instance for the entire package so we can usit throughout `mru`. Package initializer `init` deals with creating the pool.

Reply:
That's great, I was wondering if that hashpool concept could be shared throughout the whole app. :+1:

Reply:
That's great, then we can remove the pool from this package.